### PR TITLE
Led strip - sample for APA102 based strips

### PR DESCRIPTION
**Pull request by RandomReaper**
Add samples/drivers/led_strips (renamed from samples/drivers/led_ws2812) and added APA102 sample.
Just for the recap, WS2812 and APA102 are addressable LEDs, not using the same protocol, but since the drivers/led_stip provide a common interface, they can share a common sample.

Tested on nucleo_432kc with Adafruit DotStar strip.

### CODE REVIEW

File: samples/drivers/led_strip/README.rst [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
 
 Overview
 ********
 
-This sample application demonstrates basic usage of the WS2812 LED
-strip driver, for controlling LED strips using WS2812, WS2812b,
-SK6812, and compatible driver chips.
+This sample application demonstrates basic usage of a LED strip based on the
+WS2812 or APA102 controllers.
+- WS2812 is also compatible with WS2812b, SK6812, and AdaFruit NeoPixel
```
Comment (dbkinder) [inline samples/drivers/led_strip/README.rst]: need a blank line before the first item in a bullet list

**Review by dbkinder — CHANGES_REQUESTED** [review]
doc changes look good, except for one formatting fix

**Review by mbolivar — CHANGES_REQUESTED** [review]
Hello and thank you for the contribution! I appreciate having a sample for this LED strip, as I don't have one myself and wouldn't be able to write one since I can't test the driver.

However, I do not think this pull request should be merged as-is. The samples/drivers directory is about providing one sample per driver, to show just the features of that driver. Making a single sample which supports multiple drivers is not as good, because the sample applications serve as a starting out point for others to base new code on. Mixing drivers in this way means that the first thing users will have to do is delete the code for the driver their application is not using.

For example, alongside the `led_ws2812` sample there is a `led_lpd8806` sample -- one sample per driver.

So I think it would make more sense to create a new `led_apa102` directory in `samples/drivers`, which is a sample for just the APA102.

Thanks again for the contribution!

**Comment by RandomReaper** [comment]
@mbolivar, 
Thank you for your review!
I agree with you, there are some defines in that code, and it's board/led type specific. The board specific code is already in and I add some led specific one.
I think the driver documented here is the led_strip one, not the ws2812 or apa102. In the future the board and led specific code should be replaced by device tree configuration, and the final user will not even need to know which technology is behind the led strip.
In the mean time, I can provide a patch for led_strip based on apa102 (I don't want to modify/remove samples/drivers/led_apa102c/ does bit banging on a apa102 strip).

**Comment by mbolivar** [comment]
>I think the driver documented here is the led_strip one, not the ws2812 or apa102. In the future the board and led specific code should be replaced by device tree configuration, and the final user will not even need to know which technology is behind the led strip.

Ah, nice, I missed that was part of the updates that have been going on in this area. If you want to unify all the samples into one rather than having separate ones for each I'd be interested to see what that would look like. As long as the resulting code is nice and generic and it's really just DT overlay differences that'd be pretty neat. If there's going to be a bunch of ifdef this-strip else that-strip I'd say separate samples is still better.

> In the mean time, I can provide a patch for led_strip based on apa102 (I don't want to modify/remove samples/drivers/led_apa102c/ does bit banging on a apa102 strip).

Feel  free to move that to led_apa102_bitbang or so. The real driver is better.

> @mbolivar Feel free to reject this pull request, I'll redo a apa102 specific one.

No need to redo it! You can just force push an update with the new driver. That would be better actually. It will keep the history of this discussion. See step 15 in the contribution workflow:

https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow

**Comment by RandomReaper** [comment]
@mbolivar, I was wrong about the driver, led_strips currently share a common API, not a driver, so here is a separate sample for APA102 led strip.

**Review by mbolivar — CHANGES_REQUESTED** [review]
Thanks for the update! This looks 99.99% ready -- there's just one little typo to fix, and then I am +1 to merge this.  Thanks!

**Comment by RandomReaper** [comment]
Sure, looks suspicious and an exact copy of the file in ws2812! I'll fix it!

**Comment by RandomReaper** [comment]
@dbkinder, now this PR simply add a sample, so the documentation has been rewritten (so your change request is not anymore against the current documentation).
@mbolivar, no more references to the ws2812, thanks for the review

**Review by mbolivar — APPROVED** [review]

**Comment by mbolivar** [comment]
Looks good to me!

**Review by dbkinder — APPROVED** [review]
docs LGTM

**Pull request merged**